### PR TITLE
feat(cache) : allow user to reset the node selector and affinity for cached steps. Fixes #4777

### LINKED ISSUE #4777: Cached pipeline step should reset the node selector
### What steps did you take:
Create a pipeline step, and set a node selector that restricts the execution of that step only on certain nodes (in my case, on certain GPU nodes, by EC2 instance type).
Run the pipeline once, so the step is cached.
Run the pipeline again.

### What happened:
KFP recognized that the pipeline step was cached, and replaced the pod with a dummy pod that prints "This step is taken from cache.". The resource requests for that pod were reset to nothing, but the node selector was not removed. Hence, the pod still triggered autoscaling of a GPU node, waited until the GPU node was ready, was scheduled, and then immediately exited. This increases latency and wastes money as cluster-autoscaler has some latency in taking down unused nodes.

### What did you expect to happen:
The new pod has no node selector, and is scheduled anywhere, as it is a dummy pod with no effect.

### Environment:
How did you deploy Kubeflow Pipelines (KFP)?
AWS installation guide for Kubeflow.

KFP version: 1.0.0

/kind bug
/area backend

### PR DESCRIPTION
**Pull request by NikeNano**
Fixes #4777

### CODE REVIEW

File: backend/src/cache/server/mutation.go [hunk 1]
```diff
@@ -161,6 +161,20 @@ func MutatePodIfCached(req *v1beta1.AdmissionRequest, clientMgr ClientManagerInt
 			Path:  SpecContainersPath,
 			Value: dummyContainers,
 		})
+		node_restrictions, err := getEnvBool("CACHE_NODE_RESTRICTIONS")
+		if err != nil {
+			return nil, err
+		}
+		if !node_restrictions {
+			patches = append(patches, patchOperation{
```
Comment (Ark-kun) [inline backend/src/cache/server/mutation.go:169]: Perhaps you can add the patch only when the affinity is set.

**Comment by NikeNano** [comment]
will fix the test tomorrow.

File: backend/src/cache/server/mutation_test.go [hunk 2]
```diff
@@ -211,13 +218,12 @@ func TestCacheNodeRestriction(t *testing.T) {
 	executionCache := &model.ExecutionCache{
 		ExecutionCacheKey: "f5fe913be7a4516ebfe1b5de29bcb35edd12ecc776b2f33f10ca19709ea3b2f0",
 		ExecutionOutput:   "testOutput",
-		ExecutionTemplate: `{"container":{"command":["echo", "Hello"],"image":"python:3.7"}},"affinity:{"nodeAffinity":{}}"`,
+		ExecutionTemplate: `{"container": {"command": ["echo", "Hello"],"image": "python:3.7"}, "nodeSelector": {"disktype": "ssd"}}`,
```
Comment (Ark-kun) [inline backend/src/cache/server/mutation_test.go:221]: I'm not sure the spaces should be inserted. AFAIK, the hash is calculated on JSON without these spaces.

Comment (NikeNano) [reply backend/src/cache/server/mutation_test.go]: I removed the space but don't think that was the issue in this case.

File: manifests/kustomize/base/pipeline/ml-pipeline-apiserver-deployment.yaml [hunk 3]
```diff
@@ -67,7 +67,7 @@ spec:
           valueFrom:
             secretKeyRef:
               name: mlpipeline-minio-artifact
-              key: secretkey
+              key: secretkey
```
Comment (Ark-kun) [inline manifests/kustomize/base/pipeline/ml-pipeline-apiserver-deployment.yaml:70]: Is this change needed?

Comment (NikeNano) [reply manifests/kustomize/base/pipeline/ml-pipeline-apiserver-deployment.yaml:70]: No :) Misstake will clean it up.

**Comment by NikeNano** [comment]
Fixed it and removed two newlines that should not have been added @Ark-kun

**Comment by NikeNano** [comment]
I believe I might need you approval on this as well @Bobgy.

**Comment by Bobgy** [comment]
@NikeNano sorry, it's just been a busy period for me.

Thank you for the fix!

I think there are a few high level issues to discuss first:

* Does this need to be a config? Per my understanding, a dummy pod do not need any restrictions, so maybe we should enforce that. Adding unnecessary config is also technical debt.
* KFP v2 will not need this, that's another reason I don't want a config value.

**Comment by NikeNano** [comment]
> @NikeNano sorry, it's just been a busy period for me.
> 
 Completely understand!

I think we could make it opinionated and always drop the `node selection` and `affinity` since it is a dummy pod as you @Bobgy point out. @Ark-kun pointed out some use cases [here](https://github.com/kubeflow/pipelines/issues/4777#issuecomment-731527639) where you argued for the need of configurability, but maybe we could compromise and drop it. Currently see a stronger use case for not having it (sending unnecessary workloads to accelerators for dummy responses) on the pod.

**Comment by Bobgy** [comment]
Ohh, I might be missing some of his comments, let me take a look

**Comment by Bobgy** [comment]
> The situation could be the opposite: The user could have a cheap preemptible node pool where they schedule most of the tasks. But when node selector is removed, the pod is scheduled on some node pool, taking up the resources.

If we need to make a tradeoff, this situation seems less than ideal, but also acceptable. At least we know that the scheduler will put the pod on a free node, and it will finish real fast.

@Ark-kun your thoughts?

**Comment by Ark-kun** [comment]
Thank you for this improvement.
/lgtm
/approve

**Comment by Bobgy** [comment]
@Ark-kun I guess you might have missed my comments. I was worrying the new config will soon be unnecessary because of newer caching support in v2 and v2 compatible. Shall we remove the config, because it's not as useful as the maintenance cost?

**Comment by Bobgy** [comment]
@NikeNano what about your thoughts?

**Comment by NikeNano** [comment]
I feel that V2 still lacks a lot of the features from V1 and would suggest we wait a bit until we drop it. I also don't thing we should cherry pick features and drop them but instead follow some type of plan interms of support for V1 until users have migrated to V2. What is the current plan interms of support for V1 forward @Bobgy?

**Comment by Bobgy** [comment]
Yes, I totally agree. We have no plan to drop v1 untill v2 fully matures.
My main argument is that the config is not useful enough in v1 context too to be worth configurable:
https://github.com/kubeflow/pipelines/pull/5456#issuecomment-821793079

**Comment by NikeNano** [comment]
> My main argument is that the config is not useful enough in v1 context too to be worth configurable:
> [#5456 (comment)](https://github.com/kubeflow/pipelines/pull/5456#issuecomment-821793079)

Ok, get it now. I think it would be fine to default and remove the configuration in this case, I would assume that very few users would feel the need to keep the node affinity for the cached steps.
